### PR TITLE
Fix client-side password expiration detection, or remove it

### PR DESCRIPTION
**Pull request by alanking**
Addresses #38 
Addresses #48

Just putting this up for discussion purposes, for now.

This PR presents two alternatives for #38:

1. A fix for the "expiration" feature tracked by the client-side plugin to give a nicer user experience. The expiration now tracks with the TTL being used by the server, minus 1 second, by getting the grid configuration in the server-side plugin. A test has been added as well which verifies that the behavior is working as intended, and it is passing. See note below regarding the other tests.
2. The "expiration" feature has been removed. This means that `pam_interactive` would become like all of the historical iRODS authentication schemes and display an error when the PAM session has expired, forcing the user to run `iinit` again. Incidentally, the `__expire__` key is the only thing being recorded in the "state" file (other things were removed in #32), so it is not serving any purpose when the expiration feature is removed. So, I have another commit which removes the "state" file logic. All tests are passing with this option.

There is a problem with option 1 -- the fact that the client-side plugin tries to "help" conflicts with the `password_extend_lifetime` behavior (see https://docs.irods.org/4.3.2/system_overview/configuration/#configuration-overview for more information). The configuration is effectively ignored because of the additional expiration check being done in the client side plugin. So, we need to answer this question: Should the plugin honor the `password_extend_lifetime` configuration? In either case, we will need to update tests because they are not passing with option 1.

### CODE REVIEW

**Comment by trel** [comment]
Still leaning towards option 2.  Removal of code.  No competing features.  No refactoring to the client side.

Awaiting input from SURF.

**Comment by alanking** [comment]
Input received: https://github.com/irods/irods_auth_plugin_pam_interactive/issues/38#issuecomment-2298724215

Thinking on this some more, I think we should keep all of these commits just in case we need to reference a working implementation / test in the future. If we squash them, all we will see is the original implementation (which wasn't quite working for all cases, and there was no test).

File: plugin/src/main.cpp [hunk 1]
```diff
@@ -593,6 +483,49 @@ namespace irods
...
+            }
+        }
+
+        // This is the default value built into the iRODS server.
+        constexpr rodsLong_t default_password_min_time = 121;
+        const auto* option_value = output->option_value;
+        try {
+            const rodsLong_t value = std::stoll(output->option_value);
+            if (value < 0) {
+                log_pam::warn("password_min_time value in R_GRID_CONFIGURATION is invalid. value:[{}]", option_value);
```
Comment (korydraughn) [inline plugin/src/main.cpp:513]: Could include the function name?

File: plugin/src/main.cpp [hunk 2]
```diff
@@ -593,6 +483,49 @@ namespace irods
...
+            return value;
+        }
+        catch (const std::exception& e) {
+            log_pam::warn(
+                "password_min_time value in R_GRID_CONFIGURATION is invalid. value:[{}] error:[{}]",
+                option_value, e.what());
+        }
+        catch (...) {
+            log_pam::warn("password_min_time value in R_GRID_CONFIGURATION is invalid. value:[{}]", option_value);
+        }
```
Comment (korydraughn) [inline plugin/src/main.cpp:525]: Could include the function name?

**Review by korydraughn — COMMENTED** [review]
Looks good.

Do the tests pass?

**Comment by alanking** [comment]
Yes, tests are passing when the feature is removed. The tests are not all passing with the expiration feature because of the aforementioned reasons.

Interestingly... I won't be able to address the review comments without force-pushing because I would be modifying commits from before the later commits, which delete the contents of those earlier commits. :) So, just be prepared for some weird force-pushing

File: plugin/src/main.cpp (hunk 1 above)
Comment (alanking) [reply plugin/src/main.cpp]: Done

File: plugin/src/main.cpp (hunk 2 above)
Comment (alanking) [reply plugin/src/main.cpp]: Done

**Comment by alanking** [comment]
As luck would have it, I actually forgot to delete the function you commented on, so I was able to make the modifications and then delete it without any force-pushing. I made the changes and simply confirmed that the code built with the modifications before and deleting the function. I feel there is no need to re-test since the function was not being used on the latest commit, and now it does not exist at all in the latest commit.

**Comment by korydraughn** [comment]
Okay. If you're confident things are good, clean up the commits by squashing, etc. Then I'll click the approve button.

**Comment by alanking** [comment]
Squashed

**Review by korydraughn — APPROVED** [review]
Pound it.

**Comment by alanking** [comment]
#'d, mergin

**Pull request merged**